core/schema/modulesource.go

```diff
@@ -514,6 +514,106 @@ func (s *moduleSchema) filterUnInstalledDeps(ctx context.Context, bk *buildkit.C
 	return effectiveDependencies, nil
 }
 
+func (s *moduleSchema) applyDepUpdates(ctx context.Context, bk *buildkit.Client, currentDeps []*modules.ModuleConfigDependency, updateList []string, updateAll bool) ([]*modules.ModuleConfigDependency, error) {
+	if len(updateList) == 0 && !updateAll {
+		return currentDeps, nil
+	}
+
+	updatesDepsMap := map[string]parsedRefString{}
+	toBeUpdated := map[string]string{}
+	for _, dep := range updateList {
+		depParsed := parseRefString(ctx, bk, dep)
+
+		// for scenario where user tries to update using name
+		var cleanDepPath = filepath.Clean(depParsed.modPath)
+
+		updatesDepsMap[cleanDepPath] = depParsed
+		toBeUpdated[cleanDepPath] = dep
+	}
+
+	var needsUpdate = func(currentDep *modules.ModuleConfigDependency, updateDepsMap map[string]parsedRefString, updateAll bool) (*modules.ModuleConfigDependency, bool) {
+		currentDepParsed := parseRefString(ctx, bk, currentDep.Source)
+		if updateAll {
+			source := currentDepParsed.modPath
+			if currentDepParsed.hasVersion {
+				source += "@" + currentDepParsed.modVersion
+			}
+
+			return &modules.ModuleConfigDependency{
+				Name:   currentDep.Name,
+				Source: source,
+				Pin:    "",
+			}, true
+		}
+
+		for cleanDepPath, updateDepParsed := range updateDepsMap {
+			// if dep is full path then use that
+			if currentDepParsed.modPath == updateDepParsed.modPath || currentDep.Name == cleanDepPath {
+				delete(toBeUpdated, cleanDepPath)
+
+				source := currentDepParsed.modPath
+				// if a specific version was requested, use that
+				// else use whatever version current version is configured to use
+				if updateDepParsed.hasVersion {
+					source += "@" + updateDepParsed.modVersion
+				} else if currentDepParsed.hasVersion {
+					source += "@" + currentDepParsed.modVersion
+				}
+
+				return &modules.ModuleConfigDependency{
+					Name:   currentDep.Name,
+					Source: source,
+					Pin:    "",
+				}, true
+			}
+
+			if !strings.Contains(updateDepParsed.modPath, "@") {
+				continue
+			}
+
+			// for scenarios such as "dagger update <just-name>@<version>"
+			name, version, _ := strings.Cut(cleanDepPath, "@")
+			if currentDep.Name == name {
+				delete(toBeUpdated, cleanDepPath)
+
+				source := currentDepParsed.modPath
+				if version != "" {
+					source += "@" + version
+				}
+				return &modules.ModuleConfigDependency{
+					Name:   currentDep.Name,
+					Source: source,
+					Pin:    "",
+				}, true
+			}
+		}
+
+		return nil, false
+	}
+
+	updatedDependencies := []*modules.ModuleConfigDependency{}
+	for _, currentDep := range currentDeps {
+		if updatedDep, ok := needsUpdate(currentDep, updatesDepsMap, updateAll); ok {
```

Comment:
This logic with this inline function is hard to follow - there's also lots of repetition with multiple returns. We should split out an entirely separate function `applyDepUpdate` that just updates one single `ModuleConfigDependency` to an optional specified version.

I think with that, the rest of the logic becomes quite a bit clearer.

Reply:
I pulled the function out and also refactored it to make it more readable and reduce duplicate code. Please let me know if that looks better and also pls do let me know if you have more suggestions around this.